packages/flutter_tools/lib/src/build_system/phases.dart

```diff
@@ -0,0 +1,180 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:meta/meta.dart';
+
+import '../base/file_system.dart';
+import 'build_system.dart';
+import 'exceptions.dart';
+
+/// An build that produces artifacts that can be built.
+///
+/// A [BuildDefinition] can be executed via `flutter assemble`.
+class BuildDefinition {
+  /// Create a new [BuildDefinition].
+  ///
+  /// Both [name] and [phases] must not be null. Phases must not be empty
+  /// or [createBuild] will throw a [StateError].
+  const BuildDefinition({
+    @required this.name,
+    @required this.phases,
+  }) : assert(name != null && name != ''),
+       assert(phases != null);
+
+  final String name;
+  final List<BuildPhase> phases;
+
+  /// Create the root [Target] to be invoked via the [BuildSystem].
+  ///
+  /// This method is also responsible for validating the target graph, including
+  /// checking for cycles and naming collisions.
+  ///
+  /// Throws a [StateError] if [phases] is empty.
+  // TODO(jonahwilliams): add output file collision detection.
+  Future<Target> createBuild(Environment environment) async {
```

Comment:
If the there's a single `Environment` per build invocation, wdyt about injecting it?

Reply:
There is a single environment per build invocation, but not necessarily a single build invocation per tool instance.